File: services/shared/database/models/audit.py
"""
Audit log model for tracking system events and user actions.
"""

import uuid
from datetime import datetime

from sqlalchemy import Index, String, Text
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column

from .base import Base


class AuditLog(Base):
    """
    Stores audit trail for system events and user actions.

    Used for compliance, debugging, and security monitoring.
    """

    __tablename__ = "audit_logs"

    # Primary key
    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4,
    )

    # Event identification
    event_type: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
        comment="Type of event: DOCUMENT_INGESTED, QUERY_EXECUTED, etc.",
    )
    event_source: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
        comment="Service that generated the event",
    )

    # Actor information
    tenant_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        nullable=False,
        index=True,
    )
    user_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        nullable=True,
        comment="User who triggered the event, if applicable",
    )

    # Event details
    resource_type: Mapped[str | None] = mapped_column(
        String(100),
        nullable=True,
        comment="Type of resource affected: document, chunk, conversation",
    )
    resource_id: Mapped[str | None] = mapped_column(
        String(255),
        nullable=True,
        comment="ID of the affected resource",
    )
    action: Mapped[str] = mapped_column(
        String(50),
        nullable=False,
        comment="Action performed: CREATE, READ, UPDATE, DELETE, SEARCH",
    )
    description: Mapped[str | None] = mapped_column(
        Text,
        nullable=True,
        comment="Human-readable description of the event",
    )

    # Request context
    request_id: Mapped[str | None] = mapped_column(
        String(100),
        nullable=True,
        comment="Correlation ID for distributed tracing",
    )
    ip_address: Mapped[str | None] = mapped_column(
        String(45),
        nullable=True,
        comment="Client IP address (supports IPv6)",
    )
    user_agent: Mapped[str | None] = mapped_column(
        String(500),
        nullable=True,
    )

    # Additional context
    audit_metadata: Mapped[dict] = mapped_column(
        JSONB,
        default=dict,
        nullable=False,
        comment="Additional event-specific data",
    )
    audit_changes: Mapped[dict | None] = mapped_column(
        JSONB,
        nullable=True,
        comment="Before/after state for update events",
    )

    # Timestamp
    created_at: Mapped[datetime] = mapped_column(
        default=datetime.utcnow,
        nullable=False,
    )

    # Indexes
    __table_args__ = (
        Index("ix_audit_logs_tenant_created", "tenant_id", "created_at"),
        Index("ix_audit_logs_event_type", "event_type"),
        Index("ix_audit_logs_resource", "resource_type", "resource_id"),
        Index("ix_audit_logs_user", "user_id", "created_at"),
    )

    def __repr__(self) -> str:
        return f"<AuditLog(id={self.id}, event_type='{self.event_type}', action='{self.action}')>"

File: crates/rag-database/src/models/audit.rs
//! Audit log model for tracking system events and user actions.
//!
//! Used for compliance, debugging, and security monitoring.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use uuid::Uuid;

/// Audit action types.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum AuditAction {
    Create,
    Read,
    Update,
    Delete,
    Search,
    Login,
    Logout,
    Export,
    Import,
}

impl AuditAction {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Create => "CREATE",
            Self::Read => "READ",
            Self::Update => "UPDATE",
            Self::Delete => "DELETE",
            Self::Search => "SEARCH",
            Self::Login => "LOGIN",
            Self::Logout => "LOGOUT",
            Self::Export => "EXPORT",
            Self::Import => "IMPORT",
        }
    }
}

impl std::fmt::Display for AuditAction {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.as_str())
    }
}

/// Stores audit trail for system events and user actions.
///
/// Used for compliance, debugging, and security monitoring.
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct AuditLog {
    pub id: Uuid,
    pub event_type: String,
    pub event_source: String,
    pub tenant_id: Uuid,
    pub user_id: Option<Uuid>,
    pub resource_type: Option<String>,
    pub resource_id: Option<String>,
    pub action: String,
    pub description: Option<String>,
    pub request_id: Option<String>,
    pub ip_address: Option<String>,
    pub user_agent: Option<String>,
    pub audit_metadata: serde_json::Value,
    pub audit_changes: Option<serde_json::Value>,
    pub created_at: DateTime<Utc>,
}

/// Data for creating a new audit log entry.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NewAuditLog {
    pub event_type: String,
    pub event_source: String,
    pub tenant_id: Uuid,
    pub user_id: Option<Uuid>,
    pub resource_type: Option<String>,
    pub resource_id: Option<String>,
    pub action: String,
    pub description: Option<String>,
    pub request_id: Option<String>,
    pub ip_address: Option<String>,
    pub user_agent: Option<String>,
    pub audit_metadata: Option<serde_json::Value>,
    pub audit_changes: Option<serde_json::Value>,
}

impl NewAuditLog {
    /// Create a new audit log entry builder.
    pub fn builder() -> AuditLogBuilder {
        AuditLogBuilder::default()
    }
}

/// Builder for creating audit log entries.
#[derive(Debug, Default)]
pub struct AuditLogBuilder {
    event_type: Option<String>,
    event_source: Option<String>,
    tenant_id: Option<Uuid>,
    user_id: Option<Uuid>,
    resource_type: Option<String>,
    resource_id: Option<String>,
    action: Option<String>,
    description: Option<String>,
    request_id: Option<String>,
    ip_address: Option<String>,
    user_agent: Option<String>,
    audit_metadata: Option<serde_json::Value>,
    audit_changes: Option<serde_json::Value>,
}

impl AuditLogBuilder {
    pub fn event_type(mut self, event_type: impl Into<String>) -> Self {
        self.event_type = Some(event_type.into());
        self
    }

    pub fn event_source(mut self, event_source: impl Into<String>) -> Self {
        self.event_source = Some(event_source.into());
        self
    }

    pub fn tenant_id(mut self, tenant_id: Uuid) -> Self {
        self.tenant_id = Some(tenant_id);
        self
    }

    pub fn user_id(mut self, user_id: Uuid) -> Self {
        self.user_id = Some(user_id);
        self
    }

    pub fn resource(
        mut self,
        resource_type: impl Into<String>,
        resource_id: impl Into<String>,
    ) -> Self {
        self.resource_type = Some(resource_type.into());
        self.resource_id = Some(resource_id.into());
        self
    }

    pub fn action(mut self, action: AuditAction) -> Self {
        self.action = Some(action.to_string());
        self
    }

    pub fn action_str(mut self, action: impl Into<String>) -> Self {
        self.action = Some(action.into());
        self
    }

    pub fn description(mut self, description: impl Into<String>) -> Self {
        self.description = Some(description.into());
        self
    }

    pub fn request_id(mut self, request_id: impl Into<String>) -> Self {
        self.request_id = Some(request_id.into());
        self
    }

    pub fn ip_address(mut self, ip_address: impl Into<String>) -> Self {
        self.ip_address = Some(ip_address.into());
        self
    }

    pub fn user_agent(mut self, user_agent: impl Into<String>) -> Self {
        self.user_agent = Some(user_agent.into());
        self
    }

    pub fn metadata(mut self, metadata: serde_json::Value) -> Self {
        self.audit_metadata = Some(metadata);
        self
    }

    pub fn changes(mut self, changes: serde_json::Value) -> Self {
        self.audit_changes = Some(changes);
        self
    }

    /// Build the NewAuditLog, returning an error if required fields are missing.
    pub fn build(self) -> Result<NewAuditLog, &'static str> {
        Ok(NewAuditLog {
            event_type: self.event_type.ok_or("event_type is required")?,
            event_source: self.event_source.ok_or("event_source is required")?,
            tenant_id: self.tenant_id.ok_or("tenant_id is required")?,
            user_id: self.user_id,
            resource_type: self.resource_type,
            resource_id: self.resource_id,
            action: self.action.ok_or("action is required")?,
            description: self.description,
            request_id: self.request_id,
            ip_address: self.ip_address,
            user_agent: self.user_agent,
            audit_metadata: self.audit_metadata,
            audit_changes: self.audit_changes,
        })
    }
}
